[로또 1단계] 밀러 미션 제출합니다

oungsi2000 opened the pull request:
오늘 복귀하여 급하게 하느라 조금 미흡한 부분이 있을 수 있는 점 양해부탁드립니다 

지난 자동차 경주 1단계 피드백을 최대한 반영하여 구상을 해보았습니다 

잘부탁드립니다

oungsi2000 committed:
docs:README 명세 작성 완료

oungsi2000 committed:
feat:LottoService 기능 구현 완료

oungsi2000 committed:
test:View테스트 로또의 금액을 입력받는다, 로또의 금액은 1000원이며, 최대한 살 수 있는 만큼 구매한다  구현

oungsi2000 committed:
test:View테스트 당첨 로또를 입력받는다 테스트 추가 완료

oungsi2000 committed:
test:View 테스트 구현 완료

oungsi2000 committed:
fix:View 결과 화면 테스트 구현 완료

oungsi2000 committed:
test:LottoService 전체 테스트 통과

oungsi2000 committed:
refactor:package 추가

oungsi2000 committed:
refactor(LottoView):메시지 하드코딩 부분 템플릿으로 변경

oungsi2000 committed:
refactor(LottoView):require 로 로직 변경

oungsi2000 committed:
refactor:Lotto 도메인 생성

oungsi2000 committed:
fix(LottoService):중복된 보너스 번호 입력 시 버그 해결

중복된 보너스 번호 입력시 3등 -> 2등으로 표시되는 문제 해결

oungsi2000 committed:
refactor(Rank):등수 반환 로직 수정

oungsi2000 committed:
test:ktlint test

oungsi2000 committed:
refactor:하드코딩 값 상수로 고정

oungsi2000 committed:
refactor(LottoService):코틀린 함수를 이용하여 로직 수정

oungsi2000 committed:
refactor(Lotto):코틀린 컬렉션에 의존하는 메소드 내부로 숨김

oungsi2000 committed:
refactor:LottoValidator 반환 타입 변경

oungsi2000 committed:
refactor:Lotto 검증 부분 변경

oungsi2000 committed:
refacor:RankMap 일급 컬렉션 추가

malibinYun commented on src/main/kotlin/lotto/domain/Lotto.kt:
```diff
@@ -0,0 +1,21 @@
+// 이것까지 숨기는 것은 너무 나간 것일까..?
+// 그리고 애초에 생성자로 List를 받고 있는데 이걸 굳이 일급 컬렉션으로 만들 의미가 있을까?
+data class Lotto(
+    val value: List<Int>,
+) {
```
로또 번호는 1~45 이지요. Int로는 로또 번호가 아닌 것을 걸러내기는 어려워보여요.
로또 번호에 대한 값 객체를 만들어보는 건 어떨까요?

oungsi2000 replied:
일단 말리빈의 의견을 반영한 코드를 작성하였습니다

하지만 제 생각은 "아직까지는 Int가 더 좋을 것 같다" 입니다.

이유는 다음과 같습니다
1. 이미 Lotto 객체의 생성자에서 Lotto의 값 객체의 검증을 한다
2. 로또의 값 객체로 관리하게 된다면, Lotto 객체에서 한 번만 값을 생성하면 될 것을, 불필요하게 1+6번의 객체를 생성하게 된다. (리소스 낭비)
3. 테스트 코드 작성 시 코드가 늘어나게 된다
4. Lotto의 값 객체는 현재로서는 "값이 1~45 사이에 있는 것을 검증할 수 있다. " 와 "보너스 번호 플래그를 세울 수 있다" 이 두 가지의 기능만 수행합니다. 추후 값 객체의 로직이 늘어난다면 모르겠지만 현재로서는 오히려 리팩토링을 하였을 때 읽는 것이 더 크다고 생각합니다 
 
값 객체를 생성하지 않고, Int 로 구현한 커밋 해시입니다 ! 시간 여유 되시면 함께 비교해주시면 감사하겠습니다

커밋 해시 : 605b6fb1fc318ce27d14316a4935e561c66ec310
@oungsi2000
[refactor(LottoView):View와 domain간 의존관계 해결](https://github.com/woowacourse/kotlin-lotto/pull/126/commits/605b6fb1fc318ce27d14316a4935e561c66ec310) 
[605b6fb](https://github.com/woowacourse/kotlin-lotto/pull/126/commits/605b6fb1fc318ce27d14316a4935e561c66ec310)

malibinYun replied:
먼저 프로그래밍 요구사항을 확인해보시면 좋겠네요 :)

### 추가된 요구 사항 
- 모든 원시 값과 문자열을 포장한다



> 이미 Lotto 객체의 생성자에서 Lotto의 값 객체의 검증을 한다

Lotto 말고 다른 곳에서도 번호를 검증하는 곳이 생긴다면, 계속해서 그것을 추가해야한다는 사실을 다른 개발자들도 알 수 있을까요?

> 로또의 값 객체로 관리하게 된다면, Lotto 객체에서 한 번만 값을 생성하면 될 것을, 불필요하게 1+6번의 객체를 생성하게 된다. (리소스 낭비)

45개의 값 객체 인스턴스만을 만들고 재활용한다면 전혀 리소스 낭비가 생겨나지 않겠지요. 값 객체를 만든 다음에는, 인스턴스 재활용하는 것을 다음 단계로 요구 드릴 예정이에요 :)

> 테스트 코드 작성 시 코드가 늘어나게 된다

테스트 코드 작성 코드가 늘어나는 건 어떤 문제가 생기나요?
의미 없는 테스트 라인이 늘어나는 것이 확실한가요?

> Lotto의 값 객체는 현재로서는 "값이 1~45 사이에 있는 것을 검증할 수 있다. " 와 "보너스 번호 플래그를 세울 수 있다" 이 두 가지의 기능만 수행합니다. 추후 값 객체의 로직이 늘어난다면 모르겠지만 현재로서는 오히려 리팩토링을 하였을 때 읽는 것이 더 크다고 생각합니다

리팩토링을 했을 때 읽는 것이 더 크다는 게 어떤 의미인지 더 풀어서 설명해주실 수 있나요?
오히려 LottoNumber로 값이 감싸져 있기에 어떤 값을 의미하는지 코드로 충분히 표현해냈다고 생각해요 :)

oungsi2000 replied:
앗 제가 프로그래밍 요구사항을 꼼꼼히 보지 못했네요ㅠㅠ

그리고 저도 말리빈의 의견를 듣고 생각이 조금 바뀌어 내용을 공유하도록 하겠습니다

> Lotto 말고 다른 곳에서도 번호를 검증하는 곳이 생긴다면, 계속해서 그것을 추가해야한다는 사실을 다른 개발자들도 알 수 있을까요?

처음에는 Lotto의 인터페이스를 만들어 번호를 검증하는 메서드를 추가한다면 굳이 값 객체가 필요할까? 라는 생각을 했는데, Lotto 인터페이스가 아니지만, Lotto의 Number를 사용해야 하는 상황에 대해선 제가 미처 생각을 못한 것 같습니다

> 45개의 값 객체 인스턴스만을 만들고 재활용한다면 전혀 리소스 낭비가 생겨나지 않겠지요. 값 객체를 만든 다음에는, 인스턴스 재활용하는 것을 다음 단계로 요구 드릴 예정이에요 :)

이 방법도 제가 미처 생각을 못했네요 ! 45개의 인스턴스만 미리 생성하고, 정적 팩토리 메서드로 인스턴스를 반환하게끔 로직 수정했습니다 

> 테스트 코드 작성 코드가 늘어나는 건 어떤 문제가 생기나요?
의미 없는 테스트 라인이 늘어나는 것이 확실한가요?

솔직히 말리빈께 정말 죄송한 말씀일 수 있습니다만...기존에 작성한 테스트 코드를 뜯어 고치는 데에 조금 부담감을 느껴서 이런 말을 했습니다. 슬랙의 코드-질답 채널에 가짜 생성자(팩토리 메서드)를 이용하면 그닥 코드의 양이 늘어나지 않게끔 할 수 있다는 사실 또한 알았습니다.

> 리팩토링을 했을 때 읽는 것이 더 크다는 게 어떤 의미인지 더 풀어서 설명해주실 수 있나요?
오히려 LottoNumber로 값이 감싸져 있기에 어떤 값을 의미하는지 코드로 충분히 표현해냈다고 생각해요 :)

졸면서 쓰다보니 오타가 났네요ㅠㅠ읽는 것이 아니라 잃는 것입니다ㅠㅠ 기존 Int에서 LottoNumber로 리팩토링 하였을 때 '굳이?' 라는 생각이 들어 리팩토링을 하였을 때 잃는 것이 더 크다고 생각했습니다. 이 의견은 위에서 제가 생각했던 단점들을 꽤나 커버하여 이제는 나쁘지 않다고 생각하게 되었습니다

malibinYun commented on src/main/kotlin/lotto/domain/Lotto.kt:
```diff
@@ -0,0 +1,21 @@
+data class Lotto(
+    val value: List<Int>,
+) {
+    constructor(value: String) : this(requireValidLotto(value).split(",").map { it.toInt() })
+
```
입력 문자열을 파싱하는 것은, 콘솔 뷰에 굉장히 의존적이란 생각이 드네요.
안드로이드로 옮겼을 때도, 숫자 버튼을 눌렀을 때 항상 "," 로 구분되어있는 문자열이 입력으로 들어오는 것이 기대될까요?
뷰와 도메인을 분리해보는건 어떨까요?

oungsi2000 replied:
확실히 그럴 수 있겠네요 ! 의견을 반영하여 View로 해당 로직을 빼두었습니다. 

여기서 더 궁금한 점은 View 에서 domain의 의존성을 없앨 때 입니다. 
view 에서 input에 대한 검증을 끝낸 후, Controller에서 파싱 + domain 객체 생성을 하는 것이 좋다고 생각하시나요?
일단 저는 view에서 하는 일이 부족한 것 같아, String->코틀린 기본 컬렉션 으로 변환만 하고 객체 생성은 Controller에서 기본 컬렉션 -> 일급 컬렉션 으로 변환하도록 하였습니다

malibinYun replied:
어떤 객체가 역할을 맡느냐에 따라 달라지겠지요.
지금은 도메인 쪽에 더 집중해서 미션을 수행해보세요 :)

malibinYun commented on src/main/kotlin/lotto/domain/RankMap.kt:
```diff
@@ -0,0 +1,22 @@
+import lotto.global.Config.LOTTO_PRICE
+import lotto.global.Rank
+
+// 각 등수 : 나온 숫자 의 맵을 저장하는 일급컬렉션
+class RankMap(
```
RankMap이란 단어만 보고 다른 아무나 붙잡고 어떤 역할을 수행하는지 한 번에 유추해낼 수 있을까요?
이 객체가 하는 역할을 잘 생각해보고 그것을 이름에 녹여내보는 건 어떨까요?

oungsi2000 replied:
RankMap -> ScoreRankMap 으로 이름 변경하였습니다 ! 
뭔가 딱 맞아 떨어지는 이름이 생각이 안나네요..죄송합니다

malibinYun commented on src/main/kotlin/lotto/global/Config.kt:
```diff
@@ -0,0 +1,9 @@
+package lotto.global
+
+object Config {
```
상수를 한 파일에 몰아서 사용한 큰 이유가 있으신가요 ?_?

oungsi2000 replied:
상수는 추후 lotto 패키지 전역에서 사용될 수 있다고 생각되어서 한 파일에 몰아서 관리하였습니다.

malibinYun replied:
전역에서 사용되더라도 적절한 객체에 위치하면 언제든지 사용할 수 있겠지요.
적절한 객체에 해당 상수를 모두 분리해보는 건 어떨까요?

oungsi2000 replied:
처음에는 한 곳에서 몰아서 관리하는게 더 편하지 않을까? 라는 생각이었는데, 말리빈의 의견대로 각 객체에 상수를 나누는 것이 추후 값이 늘어났을 때, 조금 관리가 더 편할 것 같네요..

malibinYun commented on src/main/kotlin/lotto/global/LottoValidator.kt:
어딘가에서 이 검증기를 통과했더라도, Lotto객체 등은 검증기의 도움 없이 바로 생성할 수 있겠지요.
Lotto 객체 생성을 하고자 하는 다른 개발자는 이런 검증기가 있는지 조차 모를 수 있겠어요.
그럴 때 그렇게 만들어진 객체는 과연 신뢰도 높은 객체일까요?

oungsi2000 replied:
Lotto 객체의 생성자에서 검증기를 호출하여 비정상적인 객체의 생성을 막고 있습니다 ! 혹시 어떤 부분에서 검증기의 도움 없이 바로 생성할 수 있는지 알려주실 수 있으실까요 ? ㅠㅠ

malibinYun replied:
로또의 생성자에서 해당 메서드를 호출하고 있다는 것을 몰랐네요.
아무 네임스페이스 없이 함수를 불러 더 혼란을 일으켰나 봐요.

로또에서 하는 일인데, 따로 결합도를 높이는 것 보다, 응집도를 높여보는 건 어떨까요?

리뷰어의 혼란을, 다른 동료 개발자의 혼란이라고도 생각해보시면 좋겠어요 :)

oungsi2000 replied:
해당 부분은 네임스페이스를 추가하여 가독성을 높였습니다. 

처음에는 Lotto 객체에 requireValidLotto 메서드를 넣을까 생각했습니다. 하지만 로또의 검증 뿐 아니라, 보너스 번호, 구입 금액의 검증의 로직은 별도로 다른 객체에서 수행하기 때문에 그냥 LottoValidator에 몰아 넣었습니다.

malibinYun replied:
LottoValidator에 몰아 넣은 이유는 무엇인가요?
응집도를 높여보는 형태로 구성해보시면 좋겠네요 :)

malibinYun commented on src/test/kotlin/lottoTest/LottoServiceTest.kt:
```diff
@@ -0,0 +1,152 @@
+class LottoServiceTest {
+    private val lottoService = LottoService(Random(RANDOM_SEED))
+
+    @Test
+    @DisplayName("복권은 6개의 1부터 45까지의 중복되지 않은 랜덤 숫자로 이루어진 리스트이다")
+    fun t1() {
```
메서드 명 대신 DisplayName을 활용한 특별한 이유가 있으신가요?

oungsi2000 replied:
메서드 명으로 사용하게 된다면 백틱을 사용해야 하는데, 개인적으로는 백틱 사용이 불편하고, 또한 DisplayName으로 관리하게 된다면 여러 줄의 문자열 리터럴을 표현할 수 있어 더욱 편리하다고 생각했습니다

malibinYun commented on src/test/kotlin/lottoTest/LottoServiceTest.kt:
```diff
@@ -0,0 +1,152 @@
+    @Test
+    @DisplayName("복권은 6개의 1부터 45까지의 중복되지 않은 랜덤 숫자로 이루어진 리스트이다")
+    fun t1() {
+        // 랜덤 요소 안정성을 위하여 여러 번 반복합니다
+        repeat(1000) {
+            val lotto = lottoService.getLotto()
+            assertThat(lotto.value)
+                .hasSize(6)
+                .allMatch({ num -> num in MIN_RANDOM_NUM..MAX_RANDOM_NUM })
+                .isEqualTo(lotto.value.distinct())
+        }
```
정말 만에 하나, 10만개를 돌려도 문제가 없다고 생각했지만, 실제로 운이 좋아서 통과한 케이스라면, 이 테스트는 의미가 있을까요?
자동차 경주때 배웠던 랜덤 테스트를 잘 기억해서 다시 구성해보세요.
개발자가 완전히 제어할 수 있는 환경에서 테스트를 진행해보세요 :)

oungsi2000 replied:
확실히 그럴 수 있겠네요 ! 해당 테스트는 테스트 가능한 부분만 따로 쪼개어 LottoTest 로 이관하였습니다

malibinYun commented on src/main/kotlin/lotto/LottoService.kt:
이 객체는 굉장히 많은 일을 수행하고 있네요!
객체가 객체간에 메시지를 던져서 각자의 역할을 수행할 수 있도록 역할을 분리해보는건 어떨까요?

malibinYun replied:
대부분의 멤버 함수의 이름들이 어떤 역할을 하는지 파악하기가 매우 어려워요.
더 구체적인 행위를 묘사해보는건 어떨까요?

oungsi2000 replied:
멤버 함수 이름 또한 더 구체적으로 수정하였고, LottoService의 메서드 3개를 LottoUtil 객체로 따로 빼 두었습니다. 
개인적으로 저는 LottoService가 많은 일을 수행하고 있지 않다고 생각했는데...혹시 어떤 부분에서 많은 일을 한다고 생각하셨는지 공유해주실 수 있으실까요?

malibinYun replied:
이전 코드를 보았을 때 객체로 묶고 역할을 분리할 수 있어 보이는 로직들이 눈에 매우 많이 보였기 때문이에요 :)
현실세계의 것을 재료 삼아 객체로 만들어보세요!
Service라는 이름을 사용하지 말아보세요.

oungsi2000 replied:
LottoService -> LottoGenerator + LottoRankFinder 로 역할별로 분리하였습니다

malibinYun commented on src/main/kotlin/lotto/LottoService.kt:
```diff
@@ -0,0 +1,60 @@
+    private val random: Random,
+) {
+    fun getLotto(): Lotto {
+        // 애초에 이미 리스트를 만들고 래핑하는게 맞나? 이 로직을 lotto 내부로 숨겨야 할까?
+        val lotto = mutableListOf<Int>()
+        while (lotto.toSet().size != MAX_LOTTO_LENGTH) {
+            lotto.clear()
+            repeat(MAX_LOTTO_LENGTH) { lotto.add(random.nextInt(MIN_RANDOM_NUM, MAX_RANDOM_NUM + 1)) }
+        }
+        return Lotto(lotto.toList())
```
운이 매우 좋지 않으면 로또 하나를 생성하는데 굉장히 오래걸리겠어요.
프로그램의 성능을 운에 맡기는 것은 적절할까요?

운에 맡기지 않고도 충분히 해결할 수 있으니 while문을 삭제하고 구성해보세요!

oungsi2000 replied:
확실히 그럴 수 있겠네요 ! 해당 코드는 다음과 같이 로직을 수정하였습니다
```kotlin
fun getLotto(): Lotto {
        val lotto = mutableListOf<LottoNumber>()
        val range: LinkedList<Int> = LinkedList(IntRange(MIN_RANDOM_NUM, MAX_RANDOM_NUM).shuffled())
        repeat(MAX_LOTTO_LENGTH) { lotto.add(LottoNumber(range.poll())) }

        return Lotto(lotto.toList())
    }
```

malibinYun reviewed (changes requested):
1단계 미션 고생 많으셨어요 :)
생각해볼만한 곳에 코멘트 남겨두었으니 확인해주세요!
피드백 반영 후 다시 리뷰 요청 부탁드려요 :)

oungsi2000 committed:
fix:로또 구입금액 불일치 버그 해결

oungsi2000 committed:
refactor(LottoService):로또 생성 로직 변경
dquote>
dquote> 기존 : while문 -> distinct 로 리스트의 크기가 6개가 될 때까지 반복
dquote> 변경 : shuffled() 메소드를 이용
dquote>
dquote>

oungsi2000 committed:
refactor:LottoServiceTest 테스트 요소 변경

기존의 t1 테스트를 -> LottoTest로 이관 및 불가능한 영역의 테스트는 제외

oungsi2000 committed:
refactor(LottoService):멤버 함수 이름 변경

oungsi2000 committed:
refactor(LottoService):getLotto, getLottoRank 메서드 분리
dquote>
dquote> 이전 : LottoService에서 모두 관리
dquote> 변경 : LottoUtil 클래스로 분리
dquote>
dquote>

oungsi2000 committed:
refactor:LottoUtil 메서드 추가

oungsi2000 committed:
refactor(LottoView):View와 domain간 의존관계 해결

Controller에서 domain 객체 선택후 Service 레이어에 주입

oungsi2000 committed:
refactor:로또 번호에 대한 값 객체 생성

malibinYun commented on src/main/kotlin/lotto/domain/LottoNumber.kt:
```diff
@@ -0,0 +1,13 @@
+    val value: Int,
+) {
+    init {
+        require(value in MIN_RANDOM_NUM..MAX_RANDOM_NUM) { Message.ERR_NOT_IN_RANGE.msg }
+    }
```
로또 숫자의 범위를 나타내는 상수인데, 상수 이름만 보면 그러한 의미가 담겨있는지 전혀 알 수 없네요.
상수에 적절한 의미를 담아보는 건 어떨까요?

oungsi2000 replied:
MIN_RANDOM_NUM ->MIN_LOTTO_NUMBER
MAX_RANDOM_NUM -> MAX_LOTTO_NUMBER

로 수정했습니다 !

malibinYun commented on src/main/kotlin/lotto/domain/LottoNumber.kt:
```diff
@@ -0,0 +1,13 @@
+import lotto.global.Config.MAX_RANDOM_NUM
+import lotto.global.Config.MIN_RANDOM_NUM
+import lotto.global.Message
+
+data class LottoNumber(
+    val value: Int,
```
여러 로또를 만들 때 중복되는 번호를 생성하지 않고, 45개의 인스턴스를 재활용할 수 있도록 구성해보세요 :)

oungsi2000 replied:
```kotlin
class LottoNumber private constructor(
    val value: Int,
) {
    companion object {
        private val pool = (MIN_LOTTO_NUMBER..MAX_LOTTO_NUMBER).map { LottoNumber(it) }.associateBy { it.value }

        fun of(value: Int): LottoNumber = pool[value] ?: throw IllegalArgumentException(Message.ERR_NOT_IN_RANGE.msg)
    }
}
```

로 로직을 수정하였습니다 !

malibinYun commented on src/main/kotlin/lotto/global/LottoUtil.kt:
```diff
@@ -0,0 +1,37 @@
+import lotto.global.Config.MAX_RANDOM_NUM
+import lotto.global.Config.MIN_RANDOM_NUM
+import java.util.LinkedList
+
+object LottoUtil {
```
LottoUtil 이라는 이름만 보고 어떤 일을 수행하는지 다른 개발자들은 쉽게 알 수 있을까요?
현실세계에서 찾을 수 있을 법한 것들을 재료 삼아서 객체로 만들어보시면 좋겠어요.

이들 중 하나를 짚어 힌트를 드리면,
당첨번호 6개와 보너스 번호 1개를 어떠한 "당첨 티켓" 이라고 묶어 부를 수 있겠지요 :)

oungsi2000 replied:
LottoUtil 또한 LottoService와 병합 후 LottoGenerator + LottoRankFinder와 getRate 메소드는 ScoreRankMap으로 이관하였습니다

malibinYun reviewed (changes requested):
여러 코멘트 남겨두었으니 확인 부탁드려요!
이전 코멘트에도 여러 피드백 남겨두었습니다 :)
피드백 내용을 깊게 고민하는 시간을 가져보시면 좋겠어요.
피드백 반영 후 다시 리뷰 요청 부탁드려요!
1단계 미션은 적어도 수요일에 머지하도록 하겠습니다 :)

oungsi2000 committed:
feat:LottoNumber를 45개의 인스턴스만 생성

oungsi2000 committed:
refactor:Config 상수 분리
dquote>
dquote> 이전 : Config 파일에서 모두 관리
dquote> 변경 : Lotto 객체에서 LOTTO_LENGTH, LOTTO_PRICE, LottoNumber 객체에서 랜덤 최솟값, 최댓값 관리

oungsi2000 committed:
refactor:상수 이름 변경

MIN_RANDOM_NUM -> MAX_LOTTO_NUMBER
MAX_RANDOM_NUM -> MAX_LOTTO_NUMBER

oungsi2000 committed:
refactor:LottoValidator 함수의 네임스페이스 사용

oungsi2000 committed:
refactor:LottoService+LottoUtil->역할에 따라 분리

oungsi2000 committed:
refactor:LottoService+LottoUtil -> LottoRankFinder 로 이관

oungsi2000 committed:
refactor:LottoUtil의 getRate -> ScoreRankMap의 메소드로 이관

oungsi2000 committed:
test:ScoreRankMap 테스트 추가

oungsi2000 committed:
refactor:프로젝트 구조 변경

oungsi2000 committed:
refactor:winningLottoTicket 도메인 객체 생성

oungsi2000 committed:
refactor:LottoNumber 로직 수정

requireNonNull -> 엘비스연산자

oungsi2000 committed:
refactor:service 객체 object로 변경

oungsi2000 committed:
fix(LottoNumber):중복 상수 사용 해결

malibinYun commented on src/main/kotlin/lotto/service/LottoGenerator.kt:
```diff
@@ -0,0 +1,25 @@
+import java.util.LinkedList
+
+object LottoGenerator {
+    fun genLotto(): Lotto {
+        val lotto = mutableListOf<LottoNumber>()
+        val range: LinkedList<Int> = LinkedList(IntRange(MIN_LOTTO_NUMBER, MAX_LOTTO_NUMBER).shuffled())
+        repeat(MAX_LOTTO_LENGTH) { lotto.add(LottoNumber.of(range.poll())) }
+
+        return Lotto(lotto.toList())
+    }
```
LinkedList 없이도 충분히 구현 가능해보여요. 최대한 코틀린 코드만을 활용해보세요.

malibinYun commented on src/main/kotlin/lotto/service/LottoGenerator.kt:
```diff
@@ -0,0 +1,25 @@
+
+        return Lotto(lotto.toList())
+    }
+
+    fun genManyLotto(iterates: Int): List<Lotto> {
```
- 줄여 쓰지 않는다(축약 금지).

프로그래밍 요구사항을 꼼꼼히 확인해주세요.

malibinYun replied:
또, ManyLotto라는 메서드명은 어떤 의미인지 잘 와닿지가 않네요.
iterates라는 단어도요.
쉬운 단어들과 군더더기 없는 표현을 활용해보는 건 어떨까요?

malibinYun commented on src/main/kotlin/lotto/service/LottoRankFinder.kt:
```diff
@@ -0,0 +1,29 @@
+import lotto.domain.ScoreRankMap
+import lotto.domain.WinningLottoTicket
+
+object LottoRankFinder {
+    fun findLottoRanks(
+        manyLotto: List<Lotto>,
+        winningLottoTicket: WinningLottoTicket,
+    ): ScoreRankMap {
+        val rankMap = Rank.entries.associateWith { 0 }.toMutableMap()
+        for (lotto in manyLotto) {
+            val rank = findLottoRank(lotto, winningLottoTicket)
+            rankMap[rank] = rankMap.getOrDefault(rank, 0) + 1
+        }
+        return ScoreRankMap(rankMap.toMap())
+    }
+
+    fun findLottoRank(
+        lotto: Lotto,
+        winningLottoTicket: WinningLottoTicket,
+    ): Rank {
+        val countOfMatch = winningLottoTicket.getCountOfMatchWith(lotto)
+        val isBonusMatched = winningLottoTicket.isMatchedBonusWith(lotto)
+        return Rank.getRank(countOfMatch, isBonusMatched)
+    }
+}
```
현실세계에서 로또 티켓과 당첨티켓을 보고 결과를 계산해주는 누군가가 있을 수 있겠어요.
컴퓨터 세계에서는 현실 세계와는 달리 조금은 마법을 부릴 수 있다는 점이 큰 특징인데요,

당첨티켓에게 비교할 티켓을 넘겨주면, 그에 대한 결과를 내놓는 특이한 세계를 만들어볼 수 있겠지요.

이런 마법을 활용해서 객체가 자율성을 가지고 서로 협력하도록 구성해보는 건 어떨까요?

malibinYun commented on src/test/kotlin/lottoTest/LottoGeneratorTest.kt:
```diff
@@ -0,0 +1,30 @@
+class LottoGeneratorTest {
+    private val lottoGenerator = LottoGenerator()
+
+    @Test
+    @DisplayName("하나의 로또를 만들 수 있다")
+    fun t1() {
+        val lotto = lottoGenerator.genLotto()
+        assertThat(lotto).isInstanceOfSatisfying(Lotto::class.java) {
+            assertThat(lotto.value).hasSize(6)
+        }
+    }
```
결과의 개수만을 비교하는 것은 테스트로 의미가 전혀 없어요.
개발자가 정확하게 제어할 수 있는 환경에서, 어떤 티켓이 정확하게 반환되는지 테스트를 진행해보세요.
자동차 경주에서 배웠던 것을 활용하면 금방 적용할 수 있을거예요 :)

malibinYun commented on src/test/kotlin/lottoTest/LottoRankFinderTest.kt:
```diff
@@ -0,0 +1,132 @@
+            )
+        val winningLotto = Lotto(listOf(1, 2, 11, 10, 8, 9).map { LottoNumber.of(it) })
+        val bonus = LottoNumber.of(7)
+        val result = lottoRankFinder.findLottoRanks(manyLotto, WinningLottoTicket(winningLotto, bonus))
+        assertThat(result.map[Rank.MISS]).isEqualTo(2)
+        assertThat(result.map[Rank.FOURTH]).isEqualTo(1)
```
정확하게 예상하는 Map 객체를 만들어 서로 직접 비교해보는 건 어떨까요?

malibinYun commented on src/test/kotlin/lottoTest/LottoRankFinderTest.kt:
```diff
@@ -0,0 +1,132 @@
+    fun t1() {
+        val manyLotto =
+            listOf(
+                Lotto(listOf(1, 2, 3, 4, 5, 6).map { LottoNumber.of(it) }),
+                Lotto(listOf(1, 2, 3, 4, 5, 6).map { LottoNumber.of(it) }),
+                Lotto(listOf(1, 2, 3, 4, 5, 6).map { LottoNumber.of(it) }),
+            )
```
로또를 아래 형태로도 생성할 수 있도록 만들어보는 건 어떨까요?
```kotlin
val lotto = Lotto(1, 2, 3, 4, 5, 6)
```

malibinYun commented on src/test/kotlin/lottoTest/LottoRankFinderTest.kt:
```diff
@@ -0,0 +1,132 @@
+    @Test
+    @DisplayName("지난 주 당첨 번호와 보너스 볼을 입력받으면 복권의 당첨 등수(enum 멤버)를 반환한다")
+    fun t2_2_1() {
+        val lotto = Lotto(listOf(1, 2, 3, 4, 5, 6).map { LottoNumber.of(it) })
+        val winningLotto = Lotto(listOf(1, 2, 3, 4, 5, 7).map { LottoNumber.of(it) })
+        val bonus = LottoNumber.of(5)
+        val result = lottoRankFinder.findLottoRank(lotto, WinningLottoTicket(winningLotto, bonus))
+        assertThat(result).isEqualTo(Rank.THIRD)
+    }
+
+    @Test
+    @DisplayName("지난 주 당첨 번호와 보너스 볼을 입력받으면 복권의 당첨 등수(enum 멤버)를 반환한다")
+    fun t2_3() {
+        val lotto = Lotto(listOf(1, 2, 3, 4, 5, 6).map { LottoNumber.of(it) })
+        val winningLotto = Lotto(listOf(1, 2, 3, 4, 8, 9).map { LottoNumber.of(it) })
+        val bonus = LottoNumber.of(7)
+        val result = lottoRankFinder.findLottoRank(lotto, WinningLottoTicket(winningLotto, bonus))
+        assertThat(result).isEqualTo(Rank.FOURTH)
+    }
+
+    @Test
+    @DisplayName("지난 주 당첨 번호와 보너스 볼을 입력받으면 복권의 당첨 등수(enum 멤버)를 반환한다")
+    fun t2_4() {
+        val lotto = Lotto(listOf(1, 2, 3, 4, 5, 6).map { LottoNumber.of(it) })
+        val winningLotto = Lotto(listOf(1, 2, 3, 10, 8, 9).map { LottoNumber.of(it) })
+        val bonus = LottoNumber.of(7)
+        val result = lottoRankFinder.findLottoRank(lotto, WinningLottoTicket(winningLotto, bonus))
+        assertThat(result).isEqualTo(Rank.FIFTH)
+    }
+
+    @Test
+    @DisplayName("지난 주 당첨 번호와 보너스 볼을 입력받으면 복권의 당첨 등수(enum 멤버)를 반환한다")
+    fun t2_5() {
+        val lotto = Lotto(listOf(1, 2, 3, 4, 5, 6).map { LottoNumber.of(it) })
+        val winningLotto = Lotto(listOf(1, 2, 11, 10, 8, 9).map { LottoNumber.of(it) })
+        val bonus = LottoNumber.of(7)
+        val result = lottoRankFinder.findLottoRank(lotto, WinningLottoTicket(winningLotto, bonus))
+        assertThat(result).isEqualTo(Rank.MISS)
+    }
+}
```
테스트명이 정확히 같아서 어떤 것을 검증하고 싶은지 차이점을 보기가 어렵네요.
ParameterizedTest에 대해서도 알아보셔도 좋겠어요.

malibinYun commented on src/main/kotlin/lotto/Application.kt:
코드를 내려받아 실행해보면 아예 컴파일 에러가 발생하네요 :)

malibinYun commented on src/main/kotlin/lotto/domain/WinningLottoTicket.kt:
```diff
@@ -0,0 +1,10 @@
+data class WinningLottoTicket(
+    val lotto: Lotto,
+    val bonusNumber: LottoNumber,
+) {
+    fun getCountOfMatchWith(contrast: Lotto): Int = lotto.value.count { it in contrast.value }
```
객체의 값을 꺼내 직접 활용하기보다, 객체에게 메시지를 던져보세요 :)

malibinYun reviewed (approved):
1단계 미션 고생 많으셨어요.
다음 미션 진행을 위해 이만 머지하겠습니다.
피드백은 다음 미션에 반영해주세요 :)
